### PR TITLE
Spark 3.4: Support fanout writers in SparkPositionDeltaWrite

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds support for fanout writers in `SparkPositionDeltaWrite`.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
cc @RussellSpitzer @szehon-ho @flyrain @singhpk234 @amogh-jahagirdar @rdblue

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java [hunk 1]
```diff
@@ -130,15 +130,16 @@ private static Distribution copyOnWriteDeleteUpdateDistribution(
 
   /** Builds requirements for merge-on-read DELETE, UPDATE, MERGE operations. */
   public static SparkWriteRequirements positionDeltaRequirements(
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java:132]: There is a note at the top that this class is for internal purposes and has no API guarantees.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java [hunk 2]
```diff
@@ -389,10 +391,44 @@ protected Map<Integer, StructProjection> buildPartitionProjections(
...
+    protected PartitioningWriter<InternalRow, DataWriteResult> newDataWriter(
+        Table table, SparkFileWriterFactory writers, OutputFileFactory files, Context context) {
+
+      FileIO io = table.io();
+      boolean fanoutEnabled = context.fanoutWriterEnabled();
+      boolean inputOrdered = context.inputOrdered();
+      long targetFileSize = context.targetDataFileSize();
+
+      if (fanoutEnabled && !inputOrdered) {
+        return new FanoutDataWriter<>(writers, files, io, targetFileSize);
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:405]: No need to use the fanout writer if input is ordered. The local sort always include partition columns.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java [hunk 3]
```diff
@@ -389,10 +391,44 @@ protected Map<Integer, StructProjection> buildPartitionProjections(
...
+      } else {
+        return new ClusteredDataWriter<>(writers, files, io, targetFileSize);
+      }
+    }
+
+    // the spec requires position deletes to be ordered by file and position
+    // use a fanout writer if the input is unordered no matter whether fanout writers are enabled
+    // clustered writers only validate records for the same spec/paritition are co-located,
```
Comment (szehon-ho) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:413]: Optional: clustered writers only validate records for the same spec/paritition are co-located which is not enough for position deletes => clustered writers assume that the position deletes are already ordered by file and position?

(the validation seems a bit orthogonal to previous sentence.   I dont mind putting it in a as well somewhere, but doesnt seem its in the other comment in newDataWriter, so hence thinking its better to omit)

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java [hunk 4]
```diff
@@ -1805,23 +1805,32 @@ public void testRangeCopyOnWriteMergePartitionedSortedTable() {
   // -------------------------------------------------------------------------
   // delete mode is NOT SET -> CLUSTER BY _spec_id, _partition, _file +
   //                           LOCALLY ORDERED BY _spec_id, _partition, _file, _pos
+  // delete mode is NOT SET (fanout) -> CLUSTER BY _spec_id, _partition, _file + empty ordering
```
Comment (szehon-ho) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:1808]: Probably annoying to add another example with ORDERED BY, but is it still correct with  "ORDERED" above?    Will it not be empty_ordering?

**Review by szehon-ho — APPROVED** [review]
Looks good to me

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java (hunk 3 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:413]: I agree, let me update.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java (hunk 4 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:1808]: If there is a table sort order, we still respect it by default and fanout writer config has no effect.

Comment (szehon-ho) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:1808]: Yea I meant , in this comment as it is ORDERED & UNORDERED, fanout should not be 'empty ordering'?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java (hunk 3 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java]: Updated.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java (hunk 4 above)
Comment (szehon-ho) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkDistributionAndOrderingUtil.java:1808]: Oh I see. ok

**Review by amogh-jahagirdar — APPROVED** [review]
Looks good to me! Thanks @aokolnychyi

**Pull request merged**